File: 4 Kyu/4 Kyu - Next bigger number with the same digits.c
#include <string.h>
#include <stdlib.h>

#define BUFFSIZE 64

void swap (char *a, char *b) {

    char temp = *a;
    *a = *b;
    *b = temp;
}
unsigned long long next_bigger_number(unsigned long long entry) {

    char num[BUFFSIZE];

    sprintf (num,"%llu",entry);
    int up, pos, len = strlen (num);

    if (len <= 1) return -1;

    up = pos = len;

    while (pos-->0)
        if (num[pos] < num[pos + 1])
            break;

    if (pos == -1) return -1;

    while (up-->0)
        if (num[up] > num[pos])
            break;

    swap (&num[pos],&num[up]);

    for (int i = pos + 1; num[i] != '\0'; ++i)
        for (int j = i + 1; j < num[j] != '\0'; j++)
            if (num[i] > num[j])
                swap(&num[i],&num[j]);

    return strtoull(num,NULL, 10);
}

File: 4 Kyu/4 Kyu - Next bigger number with the same digits.rs

fn next_bigger_number(n: i64) -> i64 {

    let mut num = n.to_string().chars().collect::<Vec<_>>();
    let len = num.len();

    let mut up = len;
    let mut pos = len - 1;

    loop {
        pos -= 1;
        if num[pos] < num[pos + 1] { break }
        if pos == 0 { return -1 }
    }

    while up > 0 {
        up -= 1;
        if num[up] > num[pos] { break }
    }

    num.swap(up,pos);

    for i in pos+1 ..len {
        for j in i+1 ..len {
            if num[i] > num[j] {
                num.swap(i,j)
            }
        }
    }

    num.iter().collect::<String>().parse::<i64>().unwrap()
}
